File: javascript/703. Kth Largest Element in a Stream.js
var longestWord = function(words) {
    let wordSet = new Set(words);
    words.sort((word1, word2)=>{
        if(word1.length == word2.length){
            return word2.localeCompare(word1);
        }
        return word1.length - word2.length;
    });
    function helper (endIndex){
        
        let targetWord = words[endIndex];
        if(targetWord.length == 1){
            return targetWord;
        }
        for(let index = targetWord.length-1; index > 0; index--){
            let previousWord = targetWord.slice(0, index);
            if(wordSet.has(previousWord) == false){
                return helper(endIndex-1);
            }        
        }
        return targetWord;
    }
    return helper(words.length-1);    
};

File: javascript/703. Kth Largest Element in a Stream.java
class Solution {
    public String longestWord(String[] words) {
        
        Set <String> wordSet = new HashSet();
        Arrays.sort(words, (first, second) -> {
            if(first.length() == second.length()){ 
                return second.compareTo(first);
            }
            return first.length() - second.length();});
                                        
        for(String word: words){
            wordSet.add(word);
        }
        return compareWord(words.length-1, words, wordSet);
    }
    
    private String compareWord(int endIndex, String[] words, Set<String> wordSet){

        String targetWord = words[endIndex];
        int wordLength = targetWord.length();
        if(wordLength == 1){
            return targetWord;
        }
        for(int index = wordLength-1; index > 0; index--){
            if(wordSet.contains(targetWord.substring(0, index)) == false){
                return compareWord(endIndex-1, words, wordSet);
            }
        }
        return targetWord;
    }
}
